projects/components/src/overlay/modal/modal-service.ts

```diff
@@ -0,0 +1,146 @@
+import { Overlay } from '@angular/cdk/overlay';
+import { ComponentPortal } from '@angular/cdk/portal';
+import { Injectable, Injector } from '@angular/core';
+import { NavigationService } from '@hypertrace/common';
+import { Subscription } from 'rxjs';
+import {
+  PopoverContainerComponent,
+  PopoverContainerData,
+  POPOVER_CONTAINER_DATA
+} from '../../popover/container/popover-container.component';
+import { PopoverBackdrop } from '../../popover/popover';
+import { PopoverPositionBuilder } from '../../popover/popover-position-builder';
+import { ModalOptions, ModalOverlayConfig, ModalPosition, MODAL_DATA } from './modal';
+import { ModalOverlayComponent } from './modal-overlay.component';
+import { ModalRef } from './modal-ref';
+
+@Injectable({ providedIn: 'root' })
+export class ModalService {
+  private activeModalPopover?: ModalRef;
+  private modalCloseSubscription?: Subscription;
+
+  public constructor(
+    public readonly overlay: Overlay,
```

Comment:
like this modalref, we don't need to completely reimplement popover here. Instead, inject popoverservice - a modal is just a specialized popover.

Reply:
We need to override the ref and data injection in this class hence the choice to reimplement it here instead of injecting popoverservice.

Reply:
We don't need to override it though - we just need to provide our own injector including those things.

Reply:
I was thinking something along the lines of:
```typescript
@Injectable({
  providedIn: 'root'
})
export class ModalService {
  private activeModal?: ModalRef<unknown>;

  public constructor(private readonly popoverService: PopoverService, private readonly defaultInjector: Injector) {}

  public createModal<T>(config: ModalOverlayConfig, injector: Injector = this.defaultInjector): ModalRef<T> {
    this.activeModal?.close();

    const popover = this.popoverService.drawPopover({
      componentOrTemplate: ModalOverlayComponent,
      parentInjector: injector,
      position: {
        type: PopoverPositionType.Fixed,
        location: PopoverFixedPositionLocation.Centered
      },
      data: {
        config: config,
        parentInjector: injector
      },
      backdrop: PopoverBackdrop.Opaque
    });

    popover.closeOnNavigation();
    this.activeModal = new ModalRef(popover);

    return this.activeModal;
  }
}
```

Reply:
we could then use this data passed into the popover to create an injector for the modal that doesn't expose the underlying popover usage:

```diff
diff --git a/projects/components/src/overlay/modal/modal-overlay.component.ts b/projects/components/src/overlay/modal/modal-overlay.component.ts
index b4e8ffb..27f4dd7 100644
--- a/projects/components/src/overlay/modal/modal-overlay.component.ts
+++ b/projects/components/src/overlay/modal/modal-overlay.component.ts
@@ -1,9 +1,10 @@
-import { ChangeDetectionStrategy, Component, Inject, TemplateRef, Type } from '@angular/core';
+import { ChangeDetectionStrategy, Component, Inject, Injector, TemplateRef, Type } from '@angular/core';
 import { IconType } from '@hypertrace/assets-library';
 import { ButtonSize, ButtonStyle } from '../../button/button';
 import { POPOVER_DATA } from '../../popover/popover';
 import { PopoverRef } from '../../popover/popover-ref';
-import { ModalOverlayConfig, ModalSize } from './modal';
+import { ModalOverlayConfig, ModalSize, MODAL_DATA } from './modal';
+import { ModalRef } from './modal-ref';
 
 @Component({
   selector: 'ht-modal-overlay',
@@ -26,7 +27,7 @@ import { ModalOverlayConfig, ModalSize } from './modal';
       <div class="content-wrapper">
         <div class="content">
           <ng-container *ngIf="this.isComponentModal; else templateRenderer">
-            <ng-container *ngComponentOutlet="this.renderer"></ng-container>
+            <ng-container *ngComponentOutlet="this.renderer; injector: this.componentInjector"></ng-container>
           </ng-container>
           <ng-template #templateRenderer>
             <ng-container *ngTemplateOutlet="this.renderer"></ng-container>
@@ -42,19 +43,42 @@ export class ModalOverlayComponent {
   public readonly size: ModalSize;
   public readonly isComponentModal: boolean;
   public readonly renderer: TemplateRef<unknown> | Type<unknown>;
+  public readonly componentInjector: Injector;
 
   public visible: boolean = true;
 
-  public constructor(private readonly popoverRef: PopoverRef, @Inject(POPOVER_DATA) config: ModalOverlayConfig) {
+  public constructor(private readonly popoverRef: PopoverRef, @Inject(POPOVER_DATA) metadata: ModalMetadata) {
+    const config = metadata.config;
     this.showHeader = config.showHeader === true;
     this.modalTitle = config.title === undefined ? '' : config.title;
     this.size = config.size;
     this.isComponentModal = !(config.content instanceof TemplateRef);
     this.renderer = config.content;
+    this.componentInjector = this.buildComponentInjector(metadata, popoverRef);
   }
 
   public close(): void {
     this.visible = false;
     this.popoverRef.close();
   }
+
+  private buildComponentInjector(metadata: ModalMetadata, popoverRef: PopoverRef): Injector {
+    return Injector.create({
+      providers: [
+        {
+          provide: MODAL_DATA,
+          useValue: metadata.config.data
+        },
+        {
+          provide: ModalRef,
+          useValue: new ModalRef(popoverRef)
+        }
+      ],
+      parent:   metadata.parentInjector});
+  }
 }
+
+export interface ModalMetadata {
+  config: ModalOverlayConfig;
+  parentInjector: Injector;
+}
\ No newline at end of file

```

Reply:
I'm working with these suggestions, but I'm kind of confused where to inject the modalRef. In the changes you suggested, modalRef is created twice. One in modal service (that is returned to consumer) and the other in modalOverlay. Any recommendations? @aaron-steinfeld

Reply:
Sorry about, confused me too looking back at it, and it may need some tweaks, but here's how I got there:

The pattern we used for popover itself can be seen in popover service - first create the ref, then create the underlying component and use the result of that to initialize the ref before returning it. The one gotcha in this case is that since we're sending the ref both down to the underlying component and up to the caller, we'd want to make sure it's initialized before either receives it. Some ways we could do that:
- Have the service initialize it before returning it. In this scenario, the modaloverlay needs to know when it's "ready for use". That could be done for example by giving the modaloverlay an observable of the modalref (or even of the whole modal injector) and waiting for that before instantiating the modal. The popover would provide that once it had the popover ref back in hand.
- Give it back in an observable. This is basically the role reversal of the above, with the modaloverlay doing the initialization and at that point the service emits the ref. This would change the API of the modal service though, since it can't block waiting for the overlay to emit.
- Have two instances of modalRef. I can't really think of any cases where object equality would matter as long as there's no issues that could lead to them becoming out of sync (which I don't think there are, but worth testing - the observer invocations would likely need to be updated from my example and wrapped with checks of the closed flag to ensure complete isn't called by both ModalRefs. This is basically what I did above, saving us from having to do a back and forth between the service and overlay at the price of having two copies of the ref.

Doing a bit more in depth thinking, I think the first option is probably the best bet, but up to you. If you have questions on it, feel free to ping me.